### PR TITLE
Clarify parameter reference: null and array length

### PR DESCRIPTION
**Pull request by pvanheus**
This aims to address https://github.com/common-workflow-language/common-workflow-language/issues/761 at least on the documentation level. The associated test should be expanded to ensure that incorrect usages are not allowed (e.g. `$(null.something)`).

### CODE REVIEW

**Comment by pvanheus** [comment]
This PR updates the text of the specification and also adds tests for some incorrect usages of parameter references.

File: concepts.md [hunk 1]
```diff
@@ -510,17 +510,21 @@ or more repeats, and all other characters are literal values.
...
-     It is an error if the key does not match the required type, or the key is not found or out
-     of range.
-  7. Repeat steps 3-6
+  4. If there are no subsequent segments, terminate and return current value
+  5. Else, match the next segment
+  6. Extract the symbol, string, or index from the segment as the key
+  7. Look up the key in current value and assign as new current value.
+     1. If the key is a symbol or string, the current value must be an object.
+     2. If the key is an index, the current value must be an array or string.
+     3. If the last key is the special value 'length' the current value must be an array and the value
```
Comment (tom-tan) [inline concepts.md:523]: How about the case of `$(inputs.length)`? Users can define an input parameter `length` or a field name `length` in record types.

Comment (pvanheus) [reply concepts.md]: Good point. So this CWL works:

```
class: CommandLineTool
cwlVersion: v1.2
inputs:
  length:
    type: int
    default: 1
  bar:
    type:
      type: record
      name: bar_record
      fields:
        length:
          type: int

outputs:
  output1:
    type: int
    outputBinding:
      outputEval: $(inputs.length)
  output2:
    type: int
    outputBinding:
      outputEval: $(inputs.bar.length)
baseCommand: "true"
```
when given suitable input.

So then the logic should be: if the key has the special value 'length' and the current value is an array the value of the parameter reference is the length of the array. If the key is 'length' and the current object is not an array, normal parameter reference lookup rules apply (including the rule in step 4).

I have updated the text accordingly and added another test for these 'length on a non-array' cases.

**Comment by pvanheus** [comment]
@mr-c and @tetron: are there rules for which tags should be specified for which conformance tests?

**Comment by mr-c** [comment]
@pvanheus not written down. 

There should be one of `command_line_tool` or `workflow`

If no optional parts like `inline_javascript` are used then it must be tagged `required` as well.

The above needs to be added to the conformance doc 😂

**Review by mr-c — APPROVED** [review]

**Pull request merged**